src/main/java/io/english/service/UserAssignmentService.java

```diff
@@ -0,0 +1,104 @@
+package io.english.service;
+
+import io.english.entity.dao.*;
+import io.english.entity.request.ChangeAssignmentIsAvailableRequest;
+import io.english.entity.request.UserAnswersRequest;
+import io.english.exceptions.EntityNotFoundException;
+import io.english.exceptions.InvalidAccessException;
+import io.english.repository.UserAssignmentRepository;
+import io.english.utils.PrincipalUtils;
+import lombok.RequiredArgsConstructor;
+import org.springframework.stereotype.Service;
+
+import java.util.List;
+import java.util.Optional;
+
+@RequiredArgsConstructor
+@Service
+public class UserAssignmentService {
+    private final UserAssignmentRepository userAssignmentRepository;
+    private final AssignmentService assignmentService;
+    private final UserService userService;
+    private final AssignmentItemAnswerService assignmentItemAnswerService;
+    private final PrincipalUtils principalUtils;
+
+    public List<UserAssignment> getAvailableAssignments() {
+        Long userId = principalUtils.getUserIdFromPrincipal();
+        return getAvailableAssignments(userId);
+    }
+
+    public UserAssignment changeAssignmentIsAvailable(Long assignmentId, Long userId, ChangeAssignmentIsAvailableRequest changeAssignmentIsAvailableRequest) {
+        var assignment = assignmentService.getById(assignmentId);
+        Boolean isAvailable = changeAssignmentIsAvailableRequest.getIsAvailable();
+        var student = userService.getById(userId);
+        Optional<UserAssignment> optionalUserAssignment = userAssignmentRepository.findByAssignmentAndUser(assignment, student);
+        UserAssignment userAssignment;
+        if (optionalUserAssignment.isEmpty()) {
+            userAssignment = createUserAssignment(student, isAvailable, assignment);
+        } else {
+            userAssignment = updateUserAssignment(optionalUserAssignment.get(), isAvailable);
+        }
+        return userAssignment;
+    }
+
+    private UserAssignment createUserAssignment(User student, Boolean isAvailable, Assignment assignment) {
+        return userAssignmentRepository.save(UserAssignment.builder()
```

Comment:
Better to create UserAssignment in one operation and then save it in repository separately for better readability